### DISCUSSION TITLE
Memory leak? orm.clsregistry._ModuleMarker._remove_item()

### DISCUSSION THREAD

**Discussion by xiangh-2018** [sqlalchemy/sqlalchemy#11787] [Usage Questions]
There seems to be flaw in above mentioned method, see diff below (against latest code as of 20240826: 5b117f3d@main):

```
diff --git a/lib/sqlalchemy/orm/clsregistry.py b/lib/sqlalchemy/orm/clsregistry.py
index 26113d8b2..20db017d1 100644
--- a/lib/sqlalchemy/orm/clsregistry.py
+++ b/lib/sqlalchemy/orm/clsregistry.py
@@ -287,8 +287,9 @@ class _ModuleMarker(ClsRegistryToken):

     def _remove_item(self, name: str) -> None:
         self.contents.pop(name, None)
-        if not self.contents and self.parent is not None:
-            self.parent._remove_item(self.name)
+        if not self.contents:
+            if self.parent is not None:
+                self.parent._remove_item(self.name)
             _registries.discard(self)

     def resolve_attr(self, key: str) -> Union[_ModNS, Type[Any]]:
```

First of all, I guess this may not be a critical issue, if all mapped classes are "static", as in most cases;
But in one of our applications we use a "dynamic" pattern: basically by explicitly calling "type()" to create mapped classes, according to some "schema descriptions" from database. Since there are theoretically indefinite number of "schema descriptions", we hope these "dynamic" classes (and all related constructs) to be garbage collected asap when they are no longer needed; but instead we observed constantly increasing memory usage;
At first we found engine's default compiled_cache is holding ref's to those "dynamic" classes, direct or not; and work around by using "execution_options={'compiled_cache'=dict()}" in all select/update/insert/... operations; then this issue was found, with help of tracemalloc;
Alternative compiled_cache's come with performance penalty; and I'm afraid there can be more bumps regarding this "dynamic" pattern, as I found from an old discussion, https://sqlalchemy.narkive.com/DT3grJvl/orm-unmapping-a-class, Michael Bayer:

> all classes can be unmapped using clear_mappers(). As for unmapping and
> remapping individual classes, we don't provide a public API for this since
> the general procedure is problematic - it wouldn't work for mapped classes
> where the mappers are related to other mappers (extremely common), and
> needing to unmap/remap classes as a matter of course, except for testing
> purposes, would perform very poorly and is generally a poor pattern of
> use.

Below is some POC code:

```python

import sqlalchemy as sa
import sqlalchemy.orm as orm
from sqlalchemy.orm import clsregistry as R

import gc
import json

def create_class(name='some_class'):
    base = type(
        'Base',
        (orm.DeclarativeBase, ),
        {},
    )
    return type(
        name,
        (base, ),
        {
            '__tablename__': name,
            'rowid': orm.mapped_column(sa.Integer, primary_key=True),
        },
    )

for i in range(10):
    cls = create_class()
    del cls
    gc.collect(2)
    # XXX: orm.clsregistry._registries is supposed to be empty ?
    print(i, json.dumps(
        [
            (e.name, list(e.contents.keys()))
            for e in orm.clsregistry._registries
            if(isinstance(e, R._ModuleMarker))
        ],
        indent=4,
    ))
    #breakpoint()
```

**Comment by CaselIT** [comment]
Hi,

It seems so, `_ModuleMarker` should cleanup after itself when it has no more content, even if it does not have a parent.

Care to open a PR?

**Reply by CaselIT** [reply]
created #11788, so you can reference that issue in the PR.

thanks!

**Comment by xiangh-2018** [comment]
Hi,

Would you please do the PR thing? I'm not quite familiar with the 
process yet, :P

在 2024/8/26 18:40, Federico Caselli 写道:

**Reply by CaselIT** [reply]
Sure!

**Reply by xiangh-2018** [reply]
Thanks, :D

**Reply by xiangh-2018** [reply]
BTW, I corrected the original diff "self.parent._remove_item(self.parent.name)" to "self.parent._remove_item(self.name)" after running related unittest, :P

**Reply by CaselIT** [reply]
yes, it just needs to call `_registries.discard(self)` in all cases when empty, not only when there is a parent
